Elasticbeanstalk app boto3

sidewinder12s opened the pull request:
Upgraded the application module for boto3

sidewinder12s commented on library/elasticbeanstalk_app.py:
```diff
@@ -60,8 +60,9 @@ def describe_app(ebs, app_name):
 def list_apps(ebs, app_name):
-    apps = ebs.describe_applications(app_name)
-    apps = apps["DescribeApplicationsResponse"]["DescribeApplicationsResult"]["Applications"]
+    app_name = [app_name] if app_name is not None else None
```
I need to convert the app_name from a string to a list string for the describe_applications call. This is the only method that needs a list string. I am wondering, do you think this list_apps + describe_app to be kind of redundant? @rohitkothari

sidewinder12s replied:
I can actually fix this up later.

sidewinder12s commented:
If this looks good to you Rohit I think this is good to integrate.
👍

sidewinder12s commented:
Disregard my line comment. I understand what I have to fix here and what it is returning, but I'll do that in another pull request. Currently the module doesn't support listing all Applications if you don't specify an application name like the API supports. 

👍

rohitkothari commented on library/elasticbeanstalk_app.py:
```diff
@@ -126,29 +124,29 @@ def main():
             app = describe_app(ebs, app_name)
 
-            result = dict(changed=True, app=app)
+            result = dict(changed=True, Application=app)
```
Looks like this would change the returned dictionary field name from "app" to "Application". Are we sure we want to have this change in the returned result of module?

sidewinder12s replied:
I had figured it was more formal and closer to what the actual API returns. We are not currently using the module so this change doesn't effect us, but it might effect anyone else being able to use the new version since it changes the returned vars. Thoughts?

sidewinder12s commented on library/elasticbeanstalk_app.py:
```diff
@@ -107,13 +108,10 @@ def main():
     try:
-        ebs = boto.beanstalk.connect_to_region(region)
-
-    except boto.exception.NoAuthHandlerFound, e:
```
That is a boto specific exception. I also believe Boto3 handles this stuff natively so you don't have to.

rohitkothari replied:
Cool, I was not aware of this. :)

rohitkothari commented on library/elasticbeanstalk_app.py:
```diff
@@ -126,29 +124,29 @@ def main():
     else:
         apps = list_apps(ebs, app_name)
-        result = dict(changed=False, apps=apps)
+        result = dict(changed=False, Applications=apps)
```
If I register the output of this module, I may expect `Application` or `Applications` as a field in registered variable. Not totally sure about this, but I feel like the returned data variable should be uniform irrespective of which condition of the module gets executed.

sidewinder12s replied:
Ugh, looks like the boto3 describe_applications returns Applications, everything else returns Application. Thoughts?

sidewinder12s replied:
After checking out some of the other modules, I think i'm gonna change the return values to use the camel_dict_to_snake_dict helper function (https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/GUIDELINES.md#camel_dict_to_snake_dict)

My last Ansible module submission they had said to change my return values to do that. This will mean that it just converts camel case values from AWS to snake_case which Ansible uses (And means that it will use the dict structure/naming that Boto returns which is probably better to use than for me to define arbitrary variable names).

sidewinder12s commented:
@rohitkothari Updated to support listing all apps & to return a consistent return value. Decided to just return Applications as that is what [describe_applications()](http://boto3.readthedocs.io/en/latest/reference/services/elasticbeanstalk.html#ElasticBeanstalk.Client.describe_applications) returns. 

Thoughts?

rohitkothari commented:
The python code changes look better to me now. Also, I liked the test playbook, is there a chance that we can also paste the output of that test playbook to read through the playbook as well as output together?

Other than that, I feel it's good enough to be merged. :)

sidewinder12s commented:
```
▶ ansible-playbook elasticbeanstalk_app-tests.yml -v
Using /etc/ansible/ansible.cfg as config file
 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ***************************************************************

TASK [set_fact] ****************************************************************
ok: [localhost] => {"ansible_facts": {"app_name": "elasticbeanstalk_app-testing", "region": "us-west-2"}, "changed": false}

TASK [Ensure no app is already created] ****************************************
ok: [localhost] => {"changed": false, "output": "Application not found"}

TASK [Wait to ensure the API doesn't return stale objects] *********************
Pausing for 3 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [localhost] => {"changed": false, "delta": 3, "rc": 0, "start": "2016-11-08 23:42:26.018134", "stderr": "", "stdout": "Paused for 3.0 seconds", "stop": "2016-11-08 23:42:29.018263", "user_input": ""}

TASK [Create Application] ******************************************************
changed: [localhost] => {"Applications": {"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:29.791000+00:00"}, "changed": true}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "created_application": {
        "Applications": {
            "ApplicationName": "elasticbeanstalk_app-testing",
            "ConfigurationTemplates": [],
            "DateCreated": "2016-11-08T23:42:29.791000+00:00",
            "DateUpdated": "2016-11-08T23:42:29.791000+00:00"
        },
        "changed": true
    }
}

TASK [Test that our created application is left unchanged] *********************
ok: [localhost] => {"Applications": {"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:29.791000+00:00"}, "changed": false}

TASK [Test adding description] *************************************************
changed: [localhost] => {"Applications": {"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:34.239000+00:00", "Description": "Testing boto3 elasticbeanstalk_app Ansible Module"}, "changed": true}

TASK [Test changing the Application Description] *******************************
changed: [localhost] => {"Applications": {"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:35.581000+00:00", "Description": "Changed the description"}, "changed": true}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "updated_application": {
        "Applications": {
            "ApplicationName": "elasticbeanstalk_app-testing",
            "ConfigurationTemplates": [],
            "DateCreated": "2016-11-08T23:42:29.791000+00:00",
            "DateUpdated": "2016-11-08T23:42:35.581000+00:00",
            "Description": "Changed the description"
        },
        "changed": true
    }
}

TASK [Display created app] *****************************************************
ok: [localhost] => {"Applications": [{"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:35.581000+00:00", "Description": "Changed the description"}], "changed": false}

TASK [Display non-existant app] ************************************************
ok: [localhost] => {"changed": false, "output": "Application not found"}

TASK [List all apps] ***********************************************************
ok: [localhost] => {"Applications": [OMMITTED], "changed": false}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "all_apps": {
        "Applications": [
            {
                OMMITTED OUTPUT (was a full list)
        ],
        "changed": false
    }
}

TASK [Delete the application] **************************************************
changed: [localhost] => {"Applications": {"ApplicationName": "elasticbeanstalk_app-testing", "ConfigurationTemplates": [], "DateCreated": "2016-11-08T23:42:29.791000+00:00", "DateUpdated": "2016-11-08T23:42:35.581000+00:00", "Description": "Changed the description"}, "changed": true}

TASK [Wait to ensure the API updates] ******************************************
Pausing for 3 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [localhost] => {"changed": false, "delta": 3, "rc": 0, "start": "2016-11-08 23:42:47.689292", "stderr": "", "stdout": "Paused for 3.0 seconds", "stop": "2016-11-08 23:42:50.689898", "user_input": ""}

TASK [Ensure the deleted application is not here anymore] **********************
ok: [localhost] => {"changed": false, "output": "Application not found"}

PLAY RECAP *********************************************************************
localhost                  : ok=16   changed=4    unreachable=0    failed=0
```

sidewinder12s commented:
👍

rohitkothari commented:
I think I noticed a slight weird behavior from the output...let me explain it with example and you can help me if it's the intended behavior.

The create/update returns `Applications` as a dictionary -
```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "updated_application": {
        "Applications": {
            "ApplicationName": "elasticbeanstalk_app-testing",
            "ConfigurationTemplates": [],
            "DateCreated": "2016-11-08T23:42:29.791000+00:00",
            "DateUpdated": "2016-11-08T23:42:35.581000+00:00",
            "Description": "Changed the description"
        },
        "changed": true
    }
}
```

whereas, the list operation returns the `Applications` as a list -
```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "all_apps": {
        "Applications": [
            {
                OMMITTED OUTPUT (was a full list)
        ],
        "changed": false
    }
}
```

Wondering if this would make the output of the module look inconsistent?

sidewinder12s commented:
Fixed that inconsistency I think. 

👍

rohitkothari commented:
Great, Geoff shared the output with me in personal and it looks good and consistent now.

:+1:
